MorphicServer/Endpoint.cs

```diff
@@ -340,6 +348,122 @@ public async Task WithTransaction(Func<Database.Session, Task> operations)
                 throw new HttpError(HttpStatusCode.InternalServerError);
             }
         }
+
+        public static string? ClientIpFromRequest(HttpRequest request)
+        {
+            string? clientIp = request.Headers["x-forwarded-for"].ToString();
+            if (clientIp == null)
+            {
+                clientIp = request.Headers["x-real-ip"].ToString();
+            }
+
+            if (clientIp == null)
+            {
+                // Last resort. This is almost never correct. It's either the
+                // load-balancer in front of the server, or localhost in development.
+                clientIp = request.Headers["client_address"].ToString();
+                if (clientIp != null && IsPrivateIp(clientIp))
+                {
+                    // ignore private IP addresses
+                    clientIp = null;
+                }
+            }
+
+            return clientIp;
+        }
+        
+        private static bool IsPrivateIp(string ipAddress)
+        {
+            if(ipAddress == "::1") return true;
+            byte[] ip = IPAddress.Parse(ipAddress).GetAddressBytes();
+            switch (ip[0])
+            {
+                case 10:
+                case 127:
+                    return true;
+                case 172:
+                    return ip[1] >= 16 && ip[1] < 32;
+                case 192:
+                    return ip[1] == 168;
+                default:
+                    return false;
+            }
+        }
+
+        public class EndpointException : MorphicServerException
+        {
+            protected EndpointException(string error) : base(error)
+            {
+            }
+
+            protected EndpointException() : base()
+            {
+            }
+        }
+
+        public class NotValidPathException : EndpointException
+        {
+            public NotValidPathException(string error) : base(error)
+            {
+            }
+        }
+        public class NoServerUrlFoundException : EndpointException
+        {
+            public NoServerUrlFoundException(string error) : base(error)
+            {
+            }
+
+            public NoServerUrlFoundException() : base()
+            {
+            }
+        }
+
+        public static string GetControllerPathUrl<T>(IHeaderDictionary requestHeaders, MorphicSettings morphicSettings)
```

Comment:
Still not wild about this signature.  Go ahead and commit, but I'll probably rework it a bit.

Reply:
About the MorphicSettings? I guess I'm still learning about this dependency injection stuff.